pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go

```diff
@@ -381,31 +384,59 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context) ([]*appsv1.Deploym
 		case constants.CheckResultCreate:
 			opErr = r.client.Create(ctx, deployment)
 		case constants.CheckResultUpdate:
-			curJson, err := json.Marshal(deployment)
-			if err != nil {
-				return nil, err
+			// Check if there are any envs to remove
+			// If there, its value will be set to "delete" so we can update the patchBytes with
+			// "patch": "delete"
+			// The strategic merge patch does not remove items from list just by removing it from the patch,
+			// to delete lists items using strategic merge patch, the $patch delete pattern is used.
+			// Example:
+			// - env:
+			//   - "name": "ENV1",
+			//     "$patch": "delete"
+			for i, deploymentC := range deployment.Spec.Template.Spec.Containers {
+				envs := []corev1.EnvVar{}
+				for _, OriginalC := range originalDeployment.Spec.Template.Spec.Containers {
+					if deploymentC.Name == OriginalC.Name {
+						envsToRemove, envsToKeep := utils.CheckEnvsToRemove(deploymentC.Env, OriginalC.Env)
+						if len(envsToRemove) > 0 {
+							envs = append(envs, envsToKeep...)
+							envs = append(envs, envsToRemove...)
+						} else {
+							envs = deploymentC.Env
+						}
+					}
+				}
+				deployment.Spec.Template.Spec.Containers[i].Env = envs
 			}
 
 			// To avoid the conflict between HPA and Deployment,
 			// we need to remove the Replicas field from the deployment spec
 			// For external autoscaler, it should not remove replicas
-			modDeployment := deployment.DeepCopy()
-			if modDeployment.Annotations[constants.AutoscalerClass] != string(constants.AutoscalerClassExternal) {
-				modDeployment.Spec.Replicas = nil
+			if deployment.Annotations[constants.AutoscalerClass] != string(constants.AutoscalerClassExternal) {
+				deployment.Spec.Replicas = nil
+				originalDeployment.Spec.Replicas = nil
 			}
 
-			modJson, err := json.Marshal(modDeployment)
+			curJson, err := json.Marshal(originalDeployment)
+			if err != nil {
+				return nil, err
+			}
+			modJson, err := json.Marshal(deployment)
 			if err != nil {
 				return nil, err
 			}
+
 			// Generate the strategic merge patch between the current and modified JSON
 			patchByte, err := strategicpatch.StrategicMergePatch(curJson, modJson, appsv1.Deployment{})
 			if err != nil {
 				return nil, err
 			}
 
+			// override the envs that needs to be removed with  "$patch": "delete"
+			patchByte = []byte(strings.ReplaceAll(string(patchByte), "\"value\":\""+utils.PLACEHOLDER_FOR_DELETION+"\"", "\"$patch\":\"delete\""))
+
 			// Patch the deployment object with the strategic merge patch
-			opErr = r.client.Patch(ctx, deployment, client.RawPatch(types.StrategicMergePatchType, patchByte))
+			opErr = r.client.Patch(ctx, deployment, kclient.RawPatch(types.StrategicMergePatchType, patchByte))
```

Comment:
Not sure what was the original change to use patch, maybe we should reconsider update operation.

Reply:
Yes, I had the same question. While looking at it, debugging it a little bit, the patch bytes contains the entire deployment model, breaking the advantage of using merge patch, which allows us to patch just a tiny portion of the object instead passing the complete one.

So, in this case, using update would prevent these issues and keep the code easier.

+1 from me to move to update.